# Data Wrangling

1. Apakah ada dosen yang cenderung memberikan nilai cenderung lebih tinggi (atau lebih rendah) dibandingkan dosen lainnya?
2. Menurut data yang ada, apakah terdapat perbedaan dalam pemberian nilai antara dosen dengan Pendidikan Terakhir S3 dan S2?
3. Apakah terdapat perbedaan performa akademik antara mahasiswa yang berasal dari Madrasah/SMK dan SMU?
4. Apakah ada perbedaan yang nampak di data terhadap kemampuan akademik antara mahasiswa dan mahasiswi?
5. Apakah ada perbedaan prestasi akademik yang nampak di data antar angkatan mahasiswa?
6. Apakah ada permasalahan di data atau informasi lain yang anda temukan di data?

Menginput dan Menampilkan data 

In [30]:
import pandas as pd
df1 = pd.read_csv('C:\\Users\\MICROSOFT\\Desktop\\EVDA Lanjut\\Case Study 02 - Data Wrangling - Academics Analysis\\mengajar.csv')
df2 = pd.read_csv('C:\\Users\\MICROSOFT\\Desktop\\EVDA Lanjut\\Case Study 02 - Data Wrangling - Academics Analysis\\tb_dosen.csv')
df3 = pd.read_csv('C:\\Users\\MICROSOFT\\Desktop\\EVDA Lanjut\\Case Study 02 - Data Wrangling - Academics Analysis\\tb_kelas.csv')
df4 = pd.read_csv('C:\\Users\\MICROSOFT\\Desktop\\EVDA Lanjut\\Case Study 02 - Data Wrangling - Academics Analysis\\tb_krs.csv')
df5 = pd.read_csv('C:\\Users\\MICROSOFT\\Desktop\\EVDA Lanjut\\Case Study 02 - Data Wrangling - Academics Analysis\\tb_mahasiswa.csv')
df6 = pd.read_csv('C:\\Users\\MICROSOFT\\Desktop\\EVDA Lanjut\\Case Study 02 - Data Wrangling - Academics Analysis\\tb_mata_kuliah.csv')
df7 = pd.read_csv('C:\\Users\\MICROSOFT\\Desktop\\EVDA Lanjut\\Case Study 02 - Data Wrangling - Academics Analysis\\tb_nilai.csv')
print (df1.shape, df2.shape, df3.shape, df4.shape, df5.shape)
print (df6.shape, df7.shape, df1.shape)
df1

(14, 4) (4, 6) (4, 4) (7, 6) (15, 8)
(9, 3) (33, 7) (14, 4)


NIDN kode_MK Semester  Tahun
0    123   KAL01   ganjil   2020
1    123   KAL01   ganjil   2019
2    123   STA01    genap   2020
3    123   MAT01    genap   2020
4    123   STA02   ganjil   2019
5    234    OR01   ganjil   2019
6    234    CS01    genap   2020
7    787   KAL01   ganjil   2021
8    787   KAL01   ganjil   2022
9    787   STA02   ganjil   2021
10   345   MAT73    genap   2020
11   345   MAT53    genap   2019
12   345   STA02   ganjil   2019
13   345   KAL02   ganjil   2020

Karena untuk mengetahui jawaban dari nomer satu, maka dibutuhkan data nilai dengan data dosen, sehingga langkah selanjutnya akan ditampilkan data nilai

In [8]:
df7

nim id_mata_kuliah  NIDN  nilai Semester  Tahun     rand
0   194015          KAL01   123     70   ganjil   2019  0.50463
1   194001          KAL01   123     80   ganjil   2019      NaN
2   194002          KAL01   123     85   ganjil   2019      NaN
3   194003          KAL01   123     90   ganjil   2019      NaN
4   194004          KAL01   123     92   ganjil   2019      NaN
5   194005          KAL01   123     91   ganjil   2019      NaN
6   194006          KAL01   123     93   ganjil   2019      NaN
7   194007          KAL01   123     85   ganjil   2019      NaN
8   194008          KAL01   787     60   ganjil   2020      NaN
9   194009          KAL01   787     63   ganjil   2020      NaN
10  194010          KAL01   787     62   ganjil   2020      NaN
11  194011          KAL01   787     61   ganjil   2020      NaN
12  194012          KAL01   787     70   ganjil   2020      NaN
13  194013          KAL01   787     90   ganjil   2020      NaN
14  194014          KAL01   787     88   ganjil   2020      NaN
15  194004           CS01   234     90    genap   2020      NaN
16  194005           CS01   234     92    genap   2020      NaN
17  194006           CS01   234     85    genap   2020      NaN
18  194007           CS01   234     55    genap   2020      NaN
19  194008           CS01   234     60    genap   2020      NaN
20  194009           CS01   234     91    genap   2020      NaN
21  194010           CS01   234     90    genap   2020      NaN
22  194015          STA02   345     40   ganjil   2021      NaN
23  194002          STA02   345     50   ganjil   2021      NaN
24  194004          STA02   345     56   ganjil   2021      NaN
25  194007          STA02   345     70   ganjil   2021      NaN
26  194008          STA02   345     72   ganjil   2021      NaN
27  194011          STA02   345     31   ganjil   2021      NaN
28  194014          STA02   345     88   ganjil   2021      NaN
29  194001          KAL02   345     55   ganjil   2020      NaN
30  194002          KAL02   345     60   ganjil   2020      NaN
31  194004          KAL02   345     65   ganjil   2020      NaN
32  194007          KAL02   345     74   ganjil   2020      NaN

Dari data nilai di atas, dapat dilihat bahwa fitur yang digunakan untuk mengetahui apakah ada kecenderungan dosen memberikan nilai lebih tinggi atau lebih rendah adalah fitur NIDN yang kepanjangan dari Nomor Induk Dosen Nasional dengan fitur nilai. Maka langkah berikutnya hanya mengambil fitur NIDN dengan nilai, yang dihitung dengan menggunakan rata-rata nilai yang diberikan masing masing dosen ke mahasiswa dan menggunakan variansinya.

In [31]:
rerata_nilai = df7.groupby(df7['NIDN']).mean()[['nilai']]
rerata_nilai.rename(columns={'nilai':'rerata_nilai'},inplace=True)
rerata_nilai

rerata_nilai
NIDN              
123      85.750000
234      80.428571
345      60.090909
787      70.571429

Fungsi grupby untuk mengelompokkan data dalam kumpulan item yang sama, dari atribut  nilai dengan NIDN kita bisa mencari tahu rata-rata nilai yang sering diberikan dosen ke mahasiswa. Dari data di atas dapat diketahui kalau dosen dengan NIDN 123 rata-rata nilai terbesar yang diberikan ke mahasiswa sekitar 85 dan rata-rata nilai terendah yang diberikan ke mahasiswa adalah 60 oleh dosen dengan NIDN 345.

In [32]:
var_nilai = df7.groupby(df7['NIDN']).var()[['nilai']]
var_nilai.rename(columns={'nilai':'var_nilai'},inplace=True)
var_nilai

var_nilai
NIDN            
123    59.928571
234   252.285714
345   263.090909
787   169.285714

In [33]:
summary_nilai = pd.concat([rerata_nilai, var_nilai], axis=1) #, keys=['level1','level2'])
summary_nilai

rerata_nilai   var_nilai
NIDN                          
123      85.750000   59.928571
234      80.428571  252.285714
345      60.090909  263.090909
787      70.571429  169.285714

In [16]:
df2.head()

NIDN           nama             email Pendidikan_Terakhir  Telephone  \
0   123   Andri Basuki   anbr@googol.com                  S3   11230000   
1   234       zulfikar   dosen@email.com                  S3   22340000   
2   787           yudi    yudi@email.com                  S2   55670000   
3   345  Annisa Fitria  annniza@yyyy.com                  S2   88970000   

  Jenis_Kelamin  
0          Pria  
1          Pria  
2          Pria  
3        Wanita

In [44]:
outer joinpd.merge(summary_nilai, df2[['nama','NIDN']],right_on='NIDN', left_index=True)

rerata_nilai   var_nilai           nama  NIDN
0     85.750000   59.928571   Andri Basuki   123
1     80.428571  252.285714       zulfikar   234
3     60.090909  263.090909  Annisa Fitria   345
2     70.571429  169.285714           yudi   787

Dari tabel di atas dapat disimpulkan bahwa terdapat kecenderungan dosen dalam memberikan nilai, tampak bahwa dosen Andri basuki memiliki variansi nilainya kecil yang artinya rerata nilai 85,75 yang diberikan dosen tersebut tersebar ke banyak mahasiswa, dan dosen Annisa Fitria memiliki variansi nilai yang besar yang artinya rerata nilai 60 yang diberikan dosen tersebut hanya diberikan ke sedikit mahasiswa. 

In [50]:
pd.merge(summary_nilai, df2[['nama','Pendidikan_Terakhir','NIDN']],right_on='NIDN', left_index=True)

rerata_nilai   var_nilai           nama Pendidikan_Terakhir  NIDN
0     85.750000   59.928571   Andri Basuki                  S3   123
1     80.428571  252.285714       zulfikar                  S3   234
3     60.090909  263.090909  Annisa Fitria                  S2   345
2     70.571429  169.285714           yudi                  S2   787

In [55]:
tips = pd.merge(summary_nilai, df2[['nama','Pendidikan_Terakhir','NIDN']],right_on='NIDN', left_index=True)
tips

rerata_nilai   var_nilai           nama Pendidikan_Terakhir  NIDN
0     85.750000   59.928571   Andri Basuki                  S3   123
1     80.428571  252.285714       zulfikar                  S3   234
3     60.090909  263.090909  Annisa Fitria                  S2   345
2     70.571429  169.285714           yudi                  S2   787

In [56]:
tips.pivot_table(index=['Pendidikan_Terakhir', 'rerata_nilai'])

NIDN   var_nilai
Pendidikan_Terakhir rerata_nilai                  
S2                  60.090909      345  263.090909
                    70.571429      787  169.285714
S3                  80.428571      234  252.285714
                    85.750000      123   59.928571

In [62]:
S2 = 60.090909,70.571429
sum(S2)

130.662338

In [63]:
len(S2)

2

In [66]:
def rerata():
    return sum(S2)/len(S2)
rerata()

65.331169

In [67]:
S3 = 80.428571, 85.750000
sum(S3)

166.178571

In [68]:
len(S3)

2

In [69]:
def rerata():
    return sum(S3)/len(S3)
rerata()

83.0892855

Dari tabel di atas, dapat diketahui bahwa terdapat perbedaan dalam pemberian nilai antara dosen dengan Pendidikan Terakhir S3 dan S2, dimana dosen S2 rerata nilai yang diberikan ke beberapa mahasiswa yaitu 65 sedangkan rerata nilai yang diberikan dosen S3 yaitu 83.

In [83]:
nilairerata = df7.groupby(df7['nim']).mean()[['nilai']]
nilairerata.rename(columns={'nilai':'nilairerata'},inplace=True)
nilairerata

nilairerata
nim                
194001        67.50
194002        65.00
194003        90.00
194004        75.75
194005        91.50
194006        89.00
194007        71.00
194008        64.00
194009        77.00
194010        76.00
194011        46.00
194012        70.00
194013        90.00
194014        88.00
194015        55.00

In [84]:
var_nilai = df7.groupby(df7['nim']).var()[['nilai']]
var_nilai.rename(columns={'nilai':'var_nilai'},inplace=True)
var_nilai

var_nilai
nim              
194001     312.50
194002     325.00
194003        NaN
194004     324.25
194005       0.50
194006      32.00
194007     154.00
194008      48.00
194009     392.00
194010     392.00
194011     450.00
194012        NaN
194013        NaN
194014       0.00
194015     450.00

In [85]:
nilai_mahasiswa = pd.concat([nilairerata, var_nilai], axis=1) #, keys=['level1','level2'])
nilai_mahasiswa

nilairerata  var_nilai
nim                           
194001        67.50     312.50
194002        65.00     325.00
194003        90.00        NaN
194004        75.75     324.25
194005        91.50       0.50
194006        89.00      32.00
194007        71.00     154.00
194008        64.00      48.00
194009        77.00     392.00
194010        76.00     392.00
194011        46.00     450.00
194012        70.00        NaN
194013        90.00        NaN
194014        88.00       0.00
194015        55.00     450.00

In [87]:
df5

nim            nama                    email  Gender Asal_Sekolah  \
0   194015   Muhammad Amir       Samarinda@inet.com    Pria          SMU   
1   194001   Samsul Kurnia        Bontang@email.com    Pria          SMU   
2   194002  Bambang Arifin         Sangata@inet.com    Pria          SMU   
3   194003            Udin            Udin@inet.com    Pria           MA   
4   194004            Iwan            Iwan@inet.com    Pria           MA   
5   194005  Bambang Arifin  Bambang Arifin@inet.com    Pria          SMK   
6   194006            Wati            Wati@inet.com  Wanita          SMK   
7   194007             Ani             Ani@inet.com  Wanita          SMK   
8   194008       Rahmawati       Rahmawati@inet.com  Wanita          SMU   
9   194009            Asep            Asep@inet.com    Pria          SMU   
10  194010            Rina            rina@inet.com  Wanita           MA   
11  194011            Sari            sari@inet.com  Wanita           MA   
12  194012            Bowo            Bowo@inet.com    Pria          SMU   
13  194013            Budi            Budi@inet.com    Pria          SMK   
14  194014           Putri           putri@inet.com  Wanita          SMK   

    Telephone  Angkatan  Asal_Daerah  
0    12345678      2019  Luar Daerah  
1    12345679      2019        Lokal  
2    12345680      2019  Luar Daerah  
3    12345681      2019        Lokal  
4    12345682      2019  Luar Daerah  
5    12345683      2019        Lokal  
6    12345684      2019        Lokal  
7    12345685      2019  Luar Daerah  
8    12345686      2020        Lokal  
9    12345687      2020  Luar Daerah  
10   12345688      2020        Lokal  
11   12345689      2020        Lokal  
12   12345690      2020  Luar Daerah  
13   12345691      2020        Lokal  
14   12345692      2020        Lokal

In [94]:
mahasiswa = pd.merge(nilai_mahasiswa, df5[['Asal_Sekolah','Angkatan','nim']], on='nim', how='left')
mahasiswa

nim  nilairerata  var_nilai Asal_Sekolah  Angkatan
0   194001        67.50     312.50          SMU      2019
1   194002        65.00     325.00          SMU      2019
2   194003        90.00        NaN           MA      2019
3   194004        75.75     324.25           MA      2019
4   194005        91.50       0.50          SMK      2019
5   194006        89.00      32.00          SMK      2019
6   194007        71.00     154.00          SMK      2019
7   194008        64.00      48.00          SMU      2020
8   194009        77.00     392.00          SMU      2020
9   194010        76.00     392.00           MA      2020
10  194011        46.00     450.00           MA      2020
11  194012        70.00        NaN          SMU      2020
12  194013        90.00        NaN          SMK      2020
13  194014        88.00       0.00          SMK      2020
14  194015        55.00     450.00          SMU      2019

In [96]:
mahasiswa.pivot_table(index=['Asal_Sekolah','Angkatan','nilairerata'])

nim  var_nilai
Asal_Sekolah Angkatan nilairerata                   
MA           2019     75.75        194004     324.25
                      90.00        194003        NaN
             2020     46.00        194011     450.00
                      76.00        194010     392.00
SMK          2019     71.00        194007     154.00
                      89.00        194006      32.00
                      91.50        194005       0.50
             2020     88.00        194014       0.00
                      90.00        194013        NaN
SMU          2019     55.00        194015     450.00
                      65.00        194002     325.00
                      67.50        194001     312.50
             2020     64.00        194008      48.00
                      70.00        194012        NaN
                      77.00        194009     392.00

In [100]:
pd.crosstab(mahasiswa.Asal_Sekolah, mahasiswa.Angkatan)

Angkatan      2019  2020
Asal_Sekolah            
MA               2     2
SMK              3     2
SMU              3     3

In [98]:
MA = (75.75,90.00,46.00,76.00)
SMK = (71.00,89.00,91.50,88.00,90.00)
SMU = (55.00,65.00,67.50,64.00,70.00,77.00)

print ([sum(MA)],[sum(SMK)],[sum(SMU)])
print ([len(MA)],[len(SMK)],[len(SMU)])
def rerata(MA):
    return sum(MA)/len(MA)
def rerata(SMK):
    return sum(SMK)/len(SMK)
def rerata(SMU):
    return sum(SMU)/len(SMU)
print ([rerata(MA)],[rerata(SMK)],[rerata(SMU)])

[287.75] [429.5] [398.5]
[4] [5] [6]
[71.9375] [85.9] [66.41666666666667]


Dari data di atas dapat diketahui bahwa terdapat perbedaan performa akademik antara mahasiswa yang berasal dari Madrasah/SMK dan SMU. Mahasiswa yang memperoleh rerata nilai tinggi berasal dari SMK, dan mahasiswa yang memperoleh rerata nilai rendah berasal dari SMU. 

In [101]:
mahasiswa.drop('Asal_Sekolah',1)

C:\Users\MICROSOFT\AppData\Local\Temp\ipykernel_21492\2847023672.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  mahasiswa.drop('Asal_Sekolah',1)


nim  nilairerata  var_nilai  Angkatan
0   194001        67.50     312.50      2019
1   194002        65.00     325.00      2019
2   194003        90.00        NaN      2019
3   194004        75.75     324.25      2019
4   194005        91.50       0.50      2019
5   194006        89.00      32.00      2019
6   194007        71.00     154.00      2019
7   194008        64.00      48.00      2020
8   194009        77.00     392.00      2020
9   194010        76.00     392.00      2020
10  194011        46.00     450.00      2020
11  194012        70.00        NaN      2020
12  194013        90.00        NaN      2020
13  194014        88.00       0.00      2020
14  194015        55.00     450.00      2019

In [102]:
mahasiswa.pivot_table(index=['Angkatan','nilairerata'])

nim  var_nilai
Angkatan nilairerata                   
2019     55.00        194015     450.00
         65.00        194002     325.00
         67.50        194001     312.50
         71.00        194007     154.00
         75.75        194004     324.25
         89.00        194006      32.00
         90.00        194003        NaN
         91.50        194005       0.50
2020     46.00        194011     450.00
         64.00        194008      48.00
         70.00        194012        NaN
         76.00        194010     392.00
         77.00        194009     392.00
         88.00        194014       0.00
         90.00        194013        NaN

In [104]:
Angkatan19 = (55.00,65.00,67.50,71.00,75.75,89.00,90.00,91.50)
Angkatan20 = (46.00,64.00,70.00,76.00,77.00,88.00,90.00)

print ([sum(Angkatan19)],[sum(Angkatan20)])
print ([len(Angkatan19)],[len(Angkatan20)])
def rerata(Angkatan19):
    return sum(Angkatan19)/len(Angkatan19)
def rerata(Angkatan20):
    return sum(Angkatan20)/len(Angkatan20)
print ([rerata(Angkatan19)],[rerata(Angkatan20)])

[604.75] [511.0]
[8] [7]
[75.59375] [73.0]


Dari data di atas dapat disimpulkan bahwa terdapat perbedaan prestasi akademik yang nampak di data antar angkatan mahasiswa. Mahasiswa angkatan 2019 rata-rata mendapatkan nilai lebih tinggi dibanding mahasiswa angkatan 20.

In [107]:
pd.merge(mahasiswa, df5[['Gender','nim']], on='nim', how='left')


nim  nilairerata  var_nilai Gender_x  Angkatan Gender_y
0   194001        67.50     312.50     Pria      2019     Pria
1   194002        65.00     325.00     Pria      2019     Pria
2   194003        90.00        NaN     Pria      2019     Pria
3   194004        75.75     324.25     Pria      2019     Pria
4   194005        91.50       0.50     Pria      2019     Pria
5   194006        89.00      32.00   Wanita      2019   Wanita
6   194007        71.00     154.00   Wanita      2019   Wanita
7   194008        64.00      48.00   Wanita      2020   Wanita
8   194009        77.00     392.00     Pria      2020     Pria
9   194010        76.00     392.00   Wanita      2020   Wanita
10  194011        46.00     450.00   Wanita      2020   Wanita
11  194012        70.00        NaN     Pria      2020     Pria
12  194013        90.00        NaN     Pria      2020     Pria
13  194014        88.00       0.00   Wanita      2020   Wanita
14  194015        55.00     450.00     Pria      2019     Pria

In [110]:
Gender = mahasiswa.drop('Angkatan',1)
Gender

C:\Users\MICROSOFT\AppData\Local\Temp\ipykernel_21492\2539082552.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only.
  Gender = mahasiswa.drop('Angkatan',1)


nim  nilairerata  var_nilai  Gender
0   194001        67.50     312.50    Pria
1   194002        65.00     325.00    Pria
2   194003        90.00        NaN    Pria
3   194004        75.75     324.25    Pria
4   194005        91.50       0.50    Pria
5   194006        89.00      32.00  Wanita
6   194007        71.00     154.00  Wanita
7   194008        64.00      48.00  Wanita
8   194009        77.00     392.00    Pria
9   194010        76.00     392.00  Wanita
10  194011        46.00     450.00  Wanita
11  194012        70.00        NaN    Pria
12  194013        90.00        NaN    Pria
13  194014        88.00       0.00  Wanita
14  194015        55.00     450.00    Pria

In [111]:
Gender.pivot_table(index=['Gender','nilairerata'])

nim  var_nilai
Gender nilairerata                   
Pria   55.00        194015     450.00
       65.00        194002     325.00
       67.50        194001     312.50
       70.00        194012        NaN
       75.75        194004     324.25
       77.00        194009     392.00
       90.00        194008        NaN
       91.50        194005       0.50
Wanita 46.00        194011     450.00
       64.00        194008      48.00
       71.00        194007     154.00
       76.00        194010     392.00
       88.00        194014       0.00
       89.00        194006      32.00

In [112]:
Pria = (55.00,65.00,67.50,70.00,75.75,77.00,90.00,91.50)
Wanita = (46.00,64.00,71.00,76.00,88.00,89.00)

print ([sum(Pria)],[sum(Wanita)])
print ([len(Pria)],[len(Wanita)])
def rerata(Pria):
    return sum(Pria)/len(Wanita)
def rerata(Pria):
    return sum(Pria)/len(Wanita)
print ([rerata(Pria)],[rerata(Wanita)])

[591.75] [434.0]
[8] [6]
[98.625] [72.33333333333333]


Dari data di atas dapat disimpulkan bahwa terdapat perbedaan yang nampak di data terhadap kemampuan akademik antara mahasiswa dan mahasiswi. Rata-rata mahasiswa mendapatkan nilai lebih tinggi dibandingkan dengan mahasiswi.

In [113]:
pd.merge(df5, df7, how='inner')

nim            nama                    email  Gender Asal_Sekolah  \
0   194015   Muhammad Amir       Samarinda@inet.com    Pria          SMU   
1   194015   Muhammad Amir       Samarinda@inet.com    Pria          SMU   
2   194001   Samsul Kurnia        Bontang@email.com    Pria          SMU   
3   194001   Samsul Kurnia        Bontang@email.com    Pria          SMU   
4   194002  Bambang Arifin         Sangata@inet.com    Pria          SMU   
5   194002  Bambang Arifin         Sangata@inet.com    Pria          SMU   
6   194002  Bambang Arifin         Sangata@inet.com    Pria          SMU   
7   194003            Udin            Udin@inet.com    Pria           MA   
8   194004            Iwan            Iwan@inet.com    Pria           MA   
9   194004            Iwan            Iwan@inet.com    Pria           MA   
10  194004            Iwan            Iwan@inet.com    Pria           MA   
11  194004            Iwan            Iwan@inet.com    Pria           MA   
12  194005  Bambang Arifin  Bambang Arifin@inet.com    Pria          SMK   
13  194005  Bambang Arifin  Bambang Arifin@inet.com    Pria          SMK   
14  194006            Wati            Wati@inet.com  Wanita          SMK   
15  194006            Wati            Wati@inet.com  Wanita          SMK   
16  194007             Ani             Ani@inet.com  Wanita          SMK   
17  194007             Ani             Ani@inet.com  Wanita          SMK   
18  194007             Ani             Ani@inet.com  Wanita          SMK   
19  194007             Ani             Ani@inet.com  Wanita          SMK   
20  194008       Rahmawati       Rahmawati@inet.com  Wanita          SMU   
21  194008       Rahmawati       Rahmawati@inet.com  Wanita          SMU   
22  194008       Rahmawati       Rahmawati@inet.com  Wanita          SMU   
23  194009            Asep            Asep@inet.com    Pria          SMU   
24  194009            Asep            Asep@inet.com    Pria          SMU   
25  194010            Rina            rina@inet.com  Wanita           MA   
26  194010            Rina            rina@inet.com  Wanita           MA   
27  194011            Sari            sari@inet.com  Wanita           MA   
28  194011            Sari            sari@inet.com  Wanita           MA   
29  194012            Bowo            Bowo@inet.com    Pria          SMU   
30  194013            Budi            Budi@inet.com    Pria          SMK   
31  194014           Putri           putri@inet.com  Wanita          SMK   
32  194014           Putri           putri@inet.com  Wanita          SMK   

    Telephone  Angkatan  Asal_Daerah id_mata_kuliah  NIDN  nilai Semester  \
0    12345678      2019  Luar Daerah          KAL01   123     70   ganjil   
1    12345678      2019  Luar Daerah          STA02   345     40   ganjil   
2    12345679      2019        Lokal          KAL01   123     80   ganjil   
3    12345679      2019        Lokal          KAL02   345     55   ganjil   
4    12345680      2019  Luar Daerah          KAL01   123     85   ganjil   
5    12345680      2019  Luar Daerah          STA02   345     50   ganjil   
6    12345680      2019  Luar Daerah          KAL02   345     60   ganjil   
7    12345681      2019        Lokal          KAL01   123     90   ganjil   
8    12345682      2019  Luar Daerah          KAL01   123     92   ganjil   
9    12345682      2019  Luar Daerah           CS01   234     90    genap   
10   12345682      2019  Luar Daerah          STA02   345     56   ganjil   
11   12345682      2019  Luar Daerah          KAL02   345     65   ganjil   
12   12345683      2019        Lokal          KAL01   123     91   ganjil   
13   12345683      2019        Lokal           CS01   234     92    genap   
14   12345684      2019        Lokal          KAL01   123     93   ganjil   
15   12345684      2019        Lokal           CS01   234     85    genap   
16   12345685      2019  Luar Daerah          KAL01   123     85   ganjil   
17   12345685      2019  Luar Daerah

Dari data yang ada, ditemukan permasalahan lain di data yaitu jumlah mahasiswa yang terdapat di data nilai dengan data mahasiswa tidak sama jumlahnya.